src/AssetManagement/actions/UniswapV2SwapAction.sol

```diff
@@ -0,0 +1,136 @@
+/**
+ * Created by Arcadia Finance
+ * https://www.arcadia.finance
+ *
+ * SPDX-License-Identifier: BUSL-1.1
+ */
+pragma solidity >=0.8.0 <0.9.0;
+
+import "../ActionBase.sol";
+import "../helpers/UniswapV2Helper.sol";
+import "../utils/ActionAssetData.sol";
+import "../../interfaces/IMainRegistry.sol";
+import "../../interfaces/IVault.sol";
+import "../../interfaces/IERC20.sol";
+
+contract UniswapV2SwapAction is ActionBase, UniswapV2Helper {
+    //Maybe add mainreg address also here
+    constructor(address _router, address _mainreg) ActionBase(_mainreg) UniswapV2Helper(_router) {}
+
+    function executeAction(address _vaultAddress, address _caller, bytes calldata _actionData)
+        public
+        override
+        returns (actionAssetsData memory)
+    {
+        // decode action data
+        // (address _vaultAddress, address _caller, bytes memory _actionSpecificData) =
+        //     abi.decode(_actionData, (address, address, bytes));
+
+        //TODO Checks might be redundant?
+        require(_vaultAddress == msg.sender, "UV2_SWAP: can only be called by vaultOwner");
+        require(IVault(_vaultAddress).owner() == _caller, "UV2_SWAP: Only vaultOwner can can call this from Vault");
+
+        // preCheck data
+        (actionAssetsData memory _outgoing, actionAssetsData memory _incoming, address[] memory path) =
+            _preCheck(_vaultAddress, _actionData);
+        // execute Action
+        _execute(_vaultAddress, _outgoing, _incoming, path);
+        // postCheck data
+        (uint256[] memory _actualOutgoingAssetsAmounts, uint256[] memory _actualIncomingAssetsAmounts) =
+            _postCheck(_vaultAddress, _outgoing, _incoming);
+
+        // approve vault to transfer actionHandlers' assets.
+        for (uint256 i; i < _outgoing.assets.length; i++) {
+            IERC20(_incoming.assets[i]).approve(_vaultAddress, type(uint256).max);
+        }
+        _incoming.assetAmounts = _actualIncomingAssetsAmounts;
+
+        return (_incoming);
+    }
+
+    function _execute(
+        address _vaultAddress,
+        actionAssetsData memory _outgoing,
+        actionAssetsData memory _incoming,
+        address[] memory path
+    ) internal {
+        _uniswapV2Swap(_vaultAddress, _outgoing.assetAmounts[0], _incoming.assetAmounts[0], path);
```

Comment:
the swapExactTokensForTokens will send it to `_vaultAddress`, should go to address(this), no? How can tests for this succeed?